### DISCUSSION TITLE
Why is my Pico W picking daiseys sometimes, and why is it so slow?

### DISCUSSION THREAD

**Discussion by frohro** [earlephilhower/arduino-pico#2106] [Q&A] [answered]
I am trying to write an application that is time sensitive, and so I like to have code do what I expect.  See my example code below.   I have this very simple program to toggle GPIO18 as fast as I can.  I haven't looked at assembly language for this, but I expected it to be significantly less than 100 clock cycles to to toggle a GPIO with the CPU. but it doesn't seem that way.  The fastest I can do is 507 kHz.  But the really big thing is every once in a while there will be a 12 us pause.  See the scope photo.  
![image](https://github.com/earlephilhower/arduino-pico/assets/994565/3bcbcf21-e335-412c-bbda-440da5ae95a7)
What is happening?
```
#include <Arduino.h>

const int LED_PIN = 18;

void setup() {
    pinMode(LED_PIN, OUTPUT);
}

void loop() {
    digitalWrite(LED_PIN, HIGH);
    digitalWrite(LED_PIN, LOW);
}
```

**Comment by earlephilhower** [comment, marked as answer]
USB interrupts, possibly.  Or some other periodic IRQ.  Use `setup1` and `loop1` for tight loops since there's nothing running on that core unless you put it there.

As for the `digitalWrite` speed, it's doing a lot of work besides toggling a pin since it's a generic function and needs to check parameters/etc.  Also, after `loop` returns and before it's called again there's other housekeeping being done in the core.

For highest speed set the GPIO pin directly from inside `while(1)` loop on the 2nd core so that it never exits.
